# Оглавление

1. [Задание 1](#task1)       
2. [Задание 2](#task2)  
    2.1 [Info](#subtask1)  
    2.2 [Describe](#subtask2)  
    2.3 [Общие выводы](#subtask3)  
    2.4 [Выбросы](#subtask4)  
    2.5 [Выводы о выбросах](#subtask5)  
3. [Задание 3](#task3)  
    3.1 [Вывод о наличии пропусков](#subtask6)  
    3.2 [Гипотезы](#subtask7)  
    3.3 [Итоговый датафрей](#subtask8)  

In [1]:
import os
import pandas as pd
import numpy as np

# Задание 1. Загрузка данных
<a id='task1'></a>
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [2]:
list_columns = ['surgery?', 'Age', 'rectal_temperature', 'pulse', 'respiratory_rate', 
                'temperature_of_extremities', 'pain', 'outcome']

data_horses = pd.read_csv(os.path.join('C:', os.sep, 'Netology', 'practice_and_dz_stat', 'horse_data.csv'), header=None, 
                          na_values='?', usecols = [0, 1, 3, 4, 5, 6, 10, 22], names = list_columns)

In [3]:
data_horses.head()

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    9                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature_of_extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

Столбец Age является вторым в описании и имеет значения 1 = Adult horse и 2 = Young (< 6 months).
В Python, нумерация начинается с 0, то есть столбец Age имеет порядковый номер 1, 
при изучении содержимого столбца установлено, что в столбце содержатся значения 1 и 9, что не соответсвует значениям в описании ввиду чего, делаем вывод, что наличие значения 9 является ошибкой и соответсвует значению 2, ввиду чего производим замену 9 на 2.

In [4]:
data_horses['Age'] = data_horses['Age'].replace(9, 2)
data_horses.head()

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    2                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature_of_extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

# Задание 2. Первичное изучение данных
<a id='task2'></a>
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

### Info
<a id='subtask1'></a>

In [5]:
data_horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            242 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


### Describe
<a id='subtask2'></a>

In [6]:
print('===========================================================')
print('Основные статистики:')
data_horses.describe()

Основные статистики:


surgery?         Age  rectal_temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.080000           38.167917   71.913043   
std      0.490305    0.271746            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

In [7]:
print('===========================================================')
print('Значение моды и медианны по столбцу: ' + 'rectal temperature')
print('Моды', data_horses.rectal_temperature.round().mode()[0])
print('Медианны', data_horses.rectal_temperature.median())
print('\n')
print('===========================================================')
print('Значение моды и медианны по столбцу: ' + 'pulse')
print('Моды', data_horses.pulse.round().mode()[0])
print('Медианны', data_horses.pulse.median())
print('\n')
print('===========================================================')
print('Значение моды и медианны по столбцу: ' + 'respiratory rate')
print('Моды', data_horses.respiratory_rate.round().mode()[0])
print('Медианны', data_horses.respiratory_rate.median())
print('\n')
print('===========================================================')

Значение моды и медианны по столбцу: rectal temperature
Моды 38.0
Медианны 38.2


Значение моды и медианны по столбцу: pulse
Моды 48.0
Медианны 64.0


Значение моды и медианны по столбцу: respiratory rate
Моды 20.0
Медианны 24.5




In [8]:
# оценим распределние животных по лечению с хирургической опирацией и без нее

data_horses['surgery?'].value_counts().head()

1.0    180
2.0    119
Name: surgery?, dtype: int64

In [9]:
# оценим распределние животных по возрасту

data_horses['Age'].value_counts().head()

1    276
2     24
Name: Age, dtype: int64

In [10]:
# оценим распределние животных по температуре конечностей (крови)

data_horses['temperature_of_extremities'].value_counts().head()

3.0    109
1.0     78
2.0     30
4.0     27
Name: temperature_of_extremities, dtype: int64

In [11]:
# оценим распределние животных по уровню предполагаемого болевого синдрома животного

data_horses['pain'].value_counts().head()

3.0    67
2.0    59
5.0    42
4.0    39
1.0    38
Name: pain, dtype: int64

In [12]:
# оценим распределние животных по выживаемости

data_horses['outcome'].value_counts().head()

1.0    178
2.0     77
3.0     44
Name: outcome, dtype: int64

### Общие вывод:
<a id='subtask3'></a>
* Большенство ислледуемых животных являются взрослые особи.
* Больше половины животных проходили лечение включающее хирургическое вмешательство, была операция.
* Показатели среднего значения и моды ректальной температуры незначительно отклонены от нормы 37.8, тем не менее являются в переделах допустимого.
* Показатель среднего значения пульса, значительно выше нормы, показатель моды также превышает значение нормы 30-40, по показателю моды можем предположить, что на момент проведения исследования животные испытывали стресс и находились незначительном шоковом состоянии, тем не менее значительное отклонение среднего начения пульса от нормы говорит о наличии выброса в данных, либо присутвия в выборке больных животных испытывающих сильный болевой синдром, данные параметр требует дополнительного исследования.
* Менее у трети животных показатели температуры конечностей относятся к нормельным, также у трети животных имеелась повышенная температура конечностей, что коррелирует с ректальной температурой, у небольшого числа животных была пониженная температура конечностей, повышенний и пониженный показатель объясняются шоком и возможным болевым синдромом у животных.
* Лижь у не значительной части выборки ивотных отсутвовала тревога и боль на момент исследования, большая часть животных имела болевой синдром разной степени проявления.
* Более половины животных в выборе выжили после проведенного исследования, менее 1\6 части выборки были усыплены, менее трети от выборки животных скончалось.

### Выбросы
<a id='subtask4'></a>

In [13]:
print('===========================================================')
print('Данные по столбцу: ' + 'rectal_temperature')
q1 = data_horses['rectal_temperature'].quantile(0.25)
q3 = data_horses['rectal_temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
print('Первый квантиль (25-й процентиль)', q1)
print('Третий квантиль (75-й процентиль)', q3)
print('Межквартильный размах', iqr)
print('===========================================================')
remove_outliers = data_horses[data_horses['rectal_temperature'].between(lower_bound, upper_bound, inclusive=True)].sort_values('rectal_temperature')
print('\nВыборосы в данных:')
result_outliers = pd.concat([data_horses, remove_outliers]).drop_duplicates(keep=False)
result_outliers['rectal_temperature'].loc[result_outliers['rectal_temperature'] > 0]

Данные по столбцу: rectal_temperature
Первый квантиль (25-й процентиль) 37.8
Третий квантиль (75-й процентиль) 38.5
Межквартильный размах 0.7000000000000028

Выборосы в данных:


20     39.9
44     35.4
54     40.3
75     39.7
80     36.4
91     40.3
99     39.6
118    36.5
141    36.0
238    36.1
251    36.6
259    40.8
281    40.0
298    36.5
Name: rectal_temperature, dtype: float64

In [14]:
print('===========================================================')
print('Данные по столбцу: ' + 'pulse')
q1 = data_horses['pulse'].quantile(0.25)
q3 = data_horses['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
print('Первый квантиль (25-й процентиль)', q1)
print('Третий квантиль (75-й процентиль)', q3)
print('Межквартильный размах', iqr)
print('===========================================================')
remove_outliers = data_horses[data_horses['pulse'].between(lower_bound, upper_bound, inclusive=True)].sort_values('pulse')
print('\nВыборосы в данных:')
result_outliers = pd.concat([data_horses, remove_outliers]).drop_duplicates(keep=False)
result_outliers['pulse'].loc[result_outliers['pulse'] > 0]

Данные по столбцу: pulse
Первый квантиль (25-й процентиль) 48.0
Третий квантиль (75-й процентиль) 88.0
Межквартильный размах 40.0

Выборосы в данных:


3      164.0
41     150.0
55     160.0
255    184.0
275    150.0
Name: pulse, dtype: float64

In [15]:
print('===========================================================')
print('Данные по столбцу: ' + 'respiratory_rate')
q1 = data_horses['respiratory_rate'].quantile(0.25)
q3 = data_horses['respiratory_rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
print('Первый квантиль (25-й процентиль)', q1)
print('Третий квантиль (75-й процентиль)', q3)
print('Межквартильный размах', iqr)
print('===========================================================')
remove_outliers = data_horses[data_horses['respiratory_rate'].between(lower_bound, upper_bound, inclusive=True)].sort_values('respiratory_rate')
print('\nВыборосы в данных:')
result_outliers = pd.concat([data_horses, remove_outliers]).drop_duplicates(keep=False)
result_outliers['respiratory_rate'].loc[result_outliers['respiratory_rate'] > 0]

Данные по столбцу: respiratory_rate
Первый квантиль (25-й процентиль) 18.5
Третий квантиль (75-й процентиль) 36.0
Межквартильный размах 17.5

Выборосы в данных:


3      84.0
39     96.0
41     72.0
82     80.0
84     80.0
103    68.0
106    96.0
120    66.0
125    68.0
208    80.0
229    70.0
244    88.0
255    84.0
265    68.0
295    70.0
Name: respiratory_rate, dtype: float64

### Выводы о выбросах:
<a id='subtask5'></a>
Как видно, количество экстремальных значений незначительно от общего числа выобрки, учитывая показатели здоровья описанные в предыдущих выводах и то, что на момент исследования животные могли испытывать стресс и шок, а также иметь болевой синдром и ухудшевшееся состояния здоровья по причине болезни, делаем вывод о том, что сильно отклоняющиеся значения пульса, дыхания и ректальной температуры носят естественный характер.

# Задание 3. Работа с пропусками
<a id='task3'></a>
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [16]:
data_horses.head()

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    2                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature_of_extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

In [17]:
data_horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            242 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [18]:
print('===========================================================')
print('Пропуски в данных:')
for column, item in zip(list_columns, (data_horses.isna().mean() * 100).round(2)):
    print(f'Название столбца: {column}, наличие пропусков {item} %')
print('===========================================================')

Пропуски в данных:
Название столбца: surgery?, наличие пропусков 0.33 %
Название столбца: Age, наличие пропусков 0.0 %
Название столбца: rectal_temperature, наличие пропусков 20.0 %
Название столбца: pulse, наличие пропусков 8.0 %
Название столбца: respiratory_rate, наличие пропусков 19.33 %
Название столбца: temperature_of_extremities, наличие пропусков 18.67 %
Название столбца: pain, наличие пропусков 18.33 %
Название столбца: outcome, наличие пропусков 0.33 %


In [19]:
# предположим, мы хотим удалить только те строки, в которых как минимум 7/8 значений заполнено
update_data = data_horses.dropna(thresh=7)

In [20]:
print('===========================================================')
print('Пропуски в данных:')
for column, item in zip(list_columns, (update_data.isna().mean() * 100).round(2)):
    print(f'Название столбца: {column}, наличие пропусков {item} %')
print('===========================================================')

Пропуски в данных:
Название столбца: surgery?, наличие пропусков 0.0 %
Название столбца: Age, наличие пропусков 0.0 %
Название столбца: rectal_temperature, наличие пропусков 9.96 %
Название столбца: pulse, наличие пропусков 1.3 %
Название столбца: respiratory_rate, наличие пропусков 9.09 %
Название столбца: temperature_of_extremities, наличие пропусков 6.49 %
Название столбца: pain, наличие пропусков 6.49 %
Название столбца: outcome, наличие пропусков 0.0 %


### Вывод о наличии пропусков:
<a id='subtask6'></a>
Как видим, количество пропусков сократилось, ввиду наличия пустных строк, предположим, что наличие пропусков объясняется невозможностью обследовать животное.

In [21]:
print(f'Количество теряемых данных при исключении строк, в которых как минимум 7/8 значений заполнено: {100 - len(update_data)/len(data_horses) * 100} %')

Количество теряемых данных при исключении строк, в которых как минимум 7/8 значений заполнено: 23.0 %


Мы сохраняем 77 % данных, принимая решение о том, что это большая потеря для нашего исследования.

In [22]:
# изучим корреляцию данных

update_data.corr()

surgery?       Age  rectal_temperature     pulse  \
surgery?                    1.000000 -0.136751            0.046091 -0.244711   
Age                        -0.136751  1.000000            0.144884  0.551941   
rectal_temperature          0.046091  0.144884            1.000000  0.168683   
pulse                      -0.244711  0.551941            0.168683  1.000000   
respiratory_rate           -0.222719  0.376876            0.260113  0.447817   
temperature_of_extremities -0.096500 -0.044850            0.080454  0.304836   
pain                       -0.335052  0.066234           -0.106532  0.286517   
outcome                    -0.117550  0.048753           -0.055739  0.357030   

                            respiratory_rate  temperature_of_extremities  \
surgery?                           -0.222719                   -0.096500   
Age                                 0.376876                   -0.044850   
rectal_temperature                  0.260113                    0.080454   
pulse                               0.447817                    0.304836   
respiratory_rate                    1.000000                    0.102315   
temperature_of_extremities          0.102315                    1.000000   
pain                                0.143980                    0.201947   
outcome                             0.103425                    0.326925   

                                pain   outcome  
surgery?                   -0.335052 -0.117550  
Age                         0.066234  0.048753  
rectal_temperature         -0.106532 -0.055739  
pulse                       0.286517  0.357030  
respiratory_rate            0.143980  0.103425  
temperature_of_extremities  0.201947  0.326925  
pain                        1.000000  0.280683  
outcome                     0.280683  1.000000

### Гипотезы:
<a id='subtask7'></a>
* Делаем предположение, что показатели пусльса и частоты дыхания зависят от возрастной группы животного
* Делаем предположение, что болевой синдром зависит от была ли проведена операция животному во время лечепния
* Делаем предположение, что показатель ректальной температуры у животных носят естественный характер на момент исследования
* Делаем предположение, что показатель температуры конечностей(крови) носит естественный характер и объясняется стрессом и шоком, которые животное могло испытывать на момент исследования или как реакция на болевой синдром от течения болезни

In [23]:
fill_data_horses = data_horses

In [24]:
fill_data_horses.head()

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    2                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature_of_extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

In [25]:
# заполняем столбец пульса и частоты дыхания согласно нашей гипотезе, с учетом возрастной группы животного

fill_data_horses['pulse'].fillna(data_horses.groupby('Age')['pulse'].transform('median'), inplace=True)
fill_data_horses['respiratory_rate'].fillna(data_horses.groupby('Age')['respiratory_rate'].transform('median'), inplace=True)

In [26]:
# заполняем столбец ректальной температуры и температуры конечностей согласно нашей гипотезе

fill_data_horses['temperature_of_extremities'].fillna(fill_data_horses.temperature_of_extremities.median(), inplace=True)
fill_data_horses['rectal_temperature'].fillna(fill_data_horses.temperature_of_extremities.median(), inplace=True)

In [27]:
# в столбце о уровне боли, производим замену значения NaN на 0, для удобства дальнейше работы

fill_data_horses.pain.fillna(0, inplace=True)

In [28]:
# заполняем столбец уровня боли согласно нашей гипотезе, с учетом наличия операции в лечении животного


fill_data_horses['pain'].loc[fill_data_horses['surgery?'] == 1.].fillna(data_horses['pain'].loc[data_horses['surgery?'] == 1.].mode()[0], inplace=True)
fill_data_horses['pain'].loc[fill_data_horses['surgery?'] != 1.].fillna(data_horses['pain'].loc[data_horses['surgery?'] != 1.].mode()[0], inplace=True)

In [29]:
# оставшееся количество пустых строк не значительно, они не повлияют на наше исследования, принимаем решение об удалении 1 строки

fill_data_horses.dropna().info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         299 non-null    int64  
 2   rectal_temperature          299 non-null    float64
 3   pulse                       299 non-null    float64
 4   respiratory_rate            299 non-null    float64
 5   temperature_of_extremities  299 non-null    float64
 6   pain                        299 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 21.0 KB


### Итоговый датафрей
<a id='subtask8'></a>

In [30]:
fill_data_horses.head(15)

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
0        2.0    1                38.5   66.0              28.0   
1        1.0    1                39.2   88.0              20.0   
2        2.0    1                38.3   40.0              24.0   
3        1.0    2                39.1  164.0              84.0   
4        2.0    1                37.3  104.0              35.0   
5        2.0    1                 3.0   60.0              24.0   
6        1.0    1                37.9   48.0              16.0   
7        1.0    1                 3.0   60.0              24.0   
8        2.0    1                 3.0   80.0              36.0   
9        2.0    2                38.3   90.0              49.0   
10       1.0    1                38.1   66.0              12.0   
11       2.0    1                39.1   72.0              52.0   
12       1.0    1                37.2   42.0              12.0   
13       2.0    2                38.0   92.0              28.0   
14       1.0    1                38.2   76.0              28.0   

    temperature_of_extremities  pain  outcome  
0                          3.0   5.0      2.0  
1                          3.0   3.0      3.0  
2                          1.0   3.0      1.0  
3                          4.0   2.0      2.0  
4                          3.0   0.0      2.0  
5                          2.0   2.0      1.0  
6                          1.0   3.0      1.0  
7                          3.0   0.0      2.0  
8                          3.0   4.0      3.0  
9                          1.0   5.0      1.0  
10                         3.0   3.0      1.0  
11                         2.0   2.0      1.0  
12                         2.0   3.0      1.0  
13                         1.0   1.0      2.0  
14                         3.0   3.0      1.0